### PR TITLE
LibWeb: Begin implementing media resource seeking

### PR DESCRIPTION
**Pull request by trflynn89**
https://user-images.githubusercontent.com/5600524/231310908-ede52697-6ea4-404a-99d1-fdf9f274a420.mp4

<br />This also gives us looping:

https://user-images.githubusercontent.com/5600524/231310927-77644388-e02c-4ce0-8840-033c6e6245a9.mp4

### CODE REVIEW

**Review by Zaggy1024 — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
Comment (trflynn89) [reply Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:1052]: This spec step is more about how we are supposed to be executing web spec steps in parallel, with the ability to abort them at any time. That whole scenario is one giant FIXME, as we execute everything synchronously right now.

Comment (Zaggy1024) [reply Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:1052]: Ah, hmm, I hadn't even considered in my implementation that `PlaybackManager` could have clients calling functions from multiple threads. In that case, perhaps it is good that the logic stay here to keep things simple in the manager.

File: Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
Comment (trflynn89) [reply Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:214]: As of right now this is actually unreachable anyways. We only enter `set_duration` with a valid duration just after creating a `PlaybackManager`, and we set `m_current_playback_position` to 0 just before.

There's probably other spec steps we haven't implemented yet that would cause this to be reached, so when we get there, we'll see how it behaves I guess

File: Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
Comment (trflynn89) [reply Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:1057]: We can't assume we have a video element here, since this `HTMLMediaElement` (in the future) could also be a `HTMLAudioElement`. So tracking a boolean for the attribute seems fine.

File: Userland/Libraries/LibWeb/HTML/HTMLVideoElement.cpp
Comment (trflynn89) [reply Userland/Libraries/LibWeb/HTML/HTMLVideoElement.cpp:108]: Yeah I'll play around with it more after your PR lands, I had a waiting thing sketched out originally, so if it's best to keep that I'll add it back

File: Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
Comment (Zaggy1024) [reply Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:214]: I figured that was probably the case, no big deal for now. :^)

File: Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp
Comment (Zaggy1024) [reply Userland/Libraries/LibWeb/HTML/HTMLMediaElement.cpp:1057]: In the future, the goal is for PlaybackManager to handle both video and audio (in any combination of one or both), so it should always be there to query. If it's used for tracking multiple seek requests across threads, though, that's fine with me.

File: Userland/Libraries/LibVideo/PlaybackManager.cpp
Comment (trflynn89) [reply Userland/Libraries/LibVideo/PlaybackManager.cpp]: Indeed it is!

**Comment by trflynn89** [comment]
Updates:
* Removed LibVideo fix for looping, as that was resolved in #18300
* Removed `Video::PlaybackManager::on_end_of_stream` in favor of the improved `Video::PlaybackManager::on_playback_state_change`
* More correctly implemented the seek step which says to wait until the underlying stream has finished seeking internally

**Comment by Zaggy1024** [comment]
Sorry, thought a maintainer would want to review it now. This all looks good to me, it works quite nicely locally for me, so feel free to merge!

**Review by Zaggy1024 — APPROVED** [review]

**Pull request merged**
